dlt/common/schema/schema.py

```diff
@@ -967,42 +982,91 @@ def _verify_update_normalizers(
+                        table["name"],
+                        to_naming,
+                        from_naming,
+                        f"Attempt to rename table name to {norm_table['name']}.",
+                    )
+                # if len(norm_table["columns"]) != len(table["columns"]):
+                #     print(norm_table["columns"])
+                #     raise TableIdentifiersFrozen(
+                #         self.name,
+                #         table["name"],
+                #         to_naming,
+                #         from_naming,
+                #         "Number of columns changed after normalization. Some columns must have"
+                #         " merged.",
+                #     )
+                col_diff = set(norm_table["columns"].keys()).symmetric_difference(
+                    table["columns"].keys()
+                )
+                if len(col_diff) > 0:
+                    raise TableIdentifiersFrozen(
+                        self.name,
+                        table["name"],
+                        to_naming,
+                        from_naming,
+                        f"Some columns got renamed to {col_diff}.",
+                    )
+
+        naming_changed = from_naming and type(from_naming) is not type(to_naming)
+        if naming_changed:
             schema_tables = {}
-            for table in self._schema_tables.values():
+            # check dlt tables
+            schema_seen_data = any(
+                utils.has_table_seen_data(t) for t in self._schema_tables.values()
+            )
+            # modify dlt tables using original naming
+            orig_dlt_tables = [
+                (self.version_table_name, utils.version_table()),
+                (self.loads_table_name, utils.loads_table()),
+                (self.state_table_name, utils.pipeline_state_table(add_dlt_id=True)),
+            ]
+            for existing_table_name, original_table in orig_dlt_tables:
+                table = self._schema_tables.get(existing_table_name)
+                # state table is optional
+                if table:
+                    table = copy(table)
+                    # keep all attributes of the schema table, copy only what we need to normalize
+                    table["columns"] = original_table["columns"]
+                    norm_table = utils.normalize_table_identifiers(table, to_naming)
+                    table_seen_data = utils.has_table_seen_data(norm_table)
+                    if schema_seen_data:
+                        _verify_identifiers(table, norm_table)
+                    schema_tables[norm_table["name"]] = norm_table
+
+            schema_seen_data = False
+            for table in self.data_tables(include_incomplete=True):
+                # TODO: when lineage is fully implemented we should use source identifiers
+                # not `table` which was already normalized
                 norm_table = utils.normalize_table_identifiers(table, to_naming)
```

Comment:
Does this make sense here? Maybe I have not thought it through properly, but we don't have the original identifiers in the schema for the data tables do we? Imho there is no way to really discover wether a new naming scheme is compatible with an old one except for on the internal tables. My intuitive solution for this would have been to allow naming changes and solve the problem by explaining in the docs that you'll need to manually update your destination table names and columns if you add a new convention that ends up loading to a different table or column.

Reply:
you are right. next step in this naming convention thing is to store source identifiers in the schema (ie. attached to every table and column names). in that case whenever naming convention changes we are able to restore source identifiers and normalize them again. I do exactly this with `dlt` tables where we know the source identifiers (hardcoded in utils).

if we have source identifiers we are able to ensure that tables that have data didn't change any names when naming convention changes. currently I block changing naming conventions if there any tables with data. this is to prevent a catastrophic situation when someone changes naming by accident and tables are broken.

if you do not care of above, there's a config setting to enable that. I'm still writing the doc here:
https://dlthub.com/devel/general-usage/naming-convention#avoid-identifier-collisions